CI linux / noble clang-23 modules currently failing

In several recent pull requests, we got build failures on `linux / noble clang-23 modules` of the kind

```
FAILED: module/CMakeFiles/dealii_module_debug.dir/interface_module_partitions/lac/affine_constraints.templates.ccm.o module/CMakeFiles/dealii_module_debug.dir/dealii-interface_partition_lac_affine_constraints.templates.pcm 
/usr/bin/clang++-23 -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_SERIALIZATION_DYN_LINK -DBOOST_SERIALIZATION_NO_LIB -DDEAL_II_BUILDING_CXX20_MODULE -DDEBUG -DKOKKOS_DEPENDENCE -D_GLIBCXX_ASSERTIONS -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE -Ddealii_module_debug_EXPORTS -I/__w/dealii/dealii/build/include -I/__w/dealii/dealii/build/source -I/__w/dealii/dealii/include -I/__w/dealii/dealii/build/module/../include -isystem /__w/dealii/dealii/bundled/taskflow-3.10.0 -isystem /usr/lib/x86_64-linux-gnu/openmpi/include -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -isystem /usr/include/petsc -isystem /usr/include/suitesparse -isystem /__w/dealii/kokkos-install/include -isystem /usr/include/ArborX -isystem /usr/include/ArborX/geometry -isystem /usr/include/ArborX/cluster -isystem /usr/include/ArborX/distributed -isystem /usr/include/ArborX/interpolation -isystem /usr/include/ArborX/spatial -isystem /usr/local/include/magic_enum -isystem /usr/include/opencascade -isystem /usr/include/slepc -isystem /usr/local/include/symengine/utilities/cereal/include -isystem /usr/include/flint -isystem /usr/include/vtk-9.1 -std=c++23 -fPIC -pedantic -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-psabi -Wno-misleading-indentation -Wfloat-conversion -Qunused-arguments -Wno-unsupported-friend -Wno-pass-failed -Wno-unused-local-typedefs -openmp-simd -ffp-exception-behavior=strict -Og -ggdb -Wa,--compress-debug-sections -MD -MT module/CMakeFiles/dealii_module_debug.dir/interface_module_partitions/lac/affine_constraints.templates.ccm.o -MF module/CMakeFiles/dealii_module_debug.dir/interface_module_partitions/lac/affine_constraints.templates.ccm.o.d @module/CMakeFiles/dealii_module_debug.dir/interface_module_partitions/lac/affine_constraints.templates.ccm.o.modmap -o module/CMakeFiles/dealii_module_debug.dir/interface_module_partitions/lac/affine_constraints.templates.ccm.o -c /__w/dealii/dealii/build/module/interface_module_partitions/lac/affine_constraints.templates.ccm
In module 'dealii.external.boost' imported from /__w/dealii/dealii/build/module/interface_module_partitions/lac/affine_constraints.templates.ccm:28:
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1595:54: error: invalid operands to binary expression ('__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>' and '__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>')
 1595 |       for (_RandomAccessIterator __i = __middle; __i < __last; ++__i)
      |                                                  ~~~ ^ ~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1869:12: note: in instantiation of function template specialization 'std::__heap_select<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1869 |       std::__heap_select(__first, __middle, __last, __comp);
      |            ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1885:13: note: in instantiation of function template specialization 'std::__partial_sort<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1885 |               std::__partial_sort(__first, __last, __last, __comp);
      |                    ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1906:9: note: in instantiation of function template specialization 'std::__introsort_loop<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, long, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1906 |           std::__introsort_loop(__first, __last,
      |                ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:4841:12: note: in instantiation of function template specialization 'std::__sort<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 4841 |       std::__sort(__first, __last, __gnu_cxx::__ops::__iter_less_iter());
      |            ^
/__w/dealii/dealii/build/module/interface_module_partitions/lac/affine_constraints.templates.ccm:388:12: note: in instantiation of function template specialization 'std::sort<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>>' requested here
  388 |       std::sort(indices.begin(), indices.end());
      |            ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/system_error:318:3: note: candidate function not viable: no known conversion from '__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>' to 'const error_code' for 1st argument
  318 |   operator<=>(const error_code& __lhs, const error_code& __rhs) noexcept
      |   ^           ~~~~~~~~~~~~~~~~~~~~~~~
...
... a lot of potential matches ...
...
In module 'dealii.external.boost' imported from /__w/dealii/dealii/build/module/interface_module_partitions/lac/affine_constraints.templates.ccm:28:
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1841:18: error: invalid operands to binary expression ('__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>' and '__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>')
 1841 |           if (!(__first < __last))
      |                 ~~~~~~~ ^ ~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1858:19: note: in instantiation of function template specialization 'std::__unguarded_partition<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1858 |       return std::__unguarded_partition(__first + 1, __last, __first, __comp);
      |                   ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1890:11: note: in instantiation of function template specialization 'std::__unguarded_partition_pivot<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1890 |             std::__unguarded_partition_pivot(__first, __last, __comp);
      |                  ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:1906:9: note: in instantiation of function template specialization 'std::__introsort_loop<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, long, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 1906 |           std::__introsort_loop(__first, __last,
      |                ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/stl_algo.h:4841:12: note: in instantiation of function template specialization 'std::__sort<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>, __gnu_cxx::__ops::_Iter_less_iter>' requested here
 4841 |       std::__sort(__first, __last, __gnu_cxx::__ops::__iter_less_iter());
      |            ^
/__w/dealii/dealii/build/module/interface_module_partitions/lac/affine_constraints.templates.ccm:388:12: note: in instantiation of function template specialization 'std::sort<__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>>' requested here
  388 |       std::sort(indices.begin(), indices.end());
      |            ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/system_error:318:3: note: candidate function not viable: no known conversion from '__gnu_cxx::__normal_iterator<unsigned int *, std::vector<unsigned int>>' to 'const error_code' for 1st argument
  318 |   operator<=>(const error_code& __lhs, const error_code& __rhs) noexcept
      |   ^           ~~~~~~~~~~~~~~~~~~~~~~~
...
... a lot of potential matches ...
...
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/unique_ptr.h:893:5: note: candidate template ignored: could not match 'unique_ptr' against '__gnu_cxx::__normal_iterator'
  893 |     operator<(const unique_ptr<_Tp, _Dp>& __x,
      |     ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/unique_ptr.h:906:5: note: candidate template ignored: could not match 'unique_ptr' against '__gnu_cxx::__normal_iterator'
  906 |     operator<(const unique_ptr<_Tp, _Dp>& __x, nullptr_t)
      |     ^
/usr/lib/gcc/x86_64-linux-gnu/15/../../../../include/c++/15/bits/unique_ptr.h:916:5: note: candidate template ignored: could not match 'unique_ptr' against '__gnu_cxx::__normal_iterator'
  916 |     operator<(nullptr_t, const unique_ptr<_Tp, _Dp>& __x)
      |     ^
2 errors generated.
```

The last PR that seems to have touched this is #20003. But that PR had the CI passing, so I am not sure whether that's the right place. I'd appreciate someone with more knowledge on these aspects to look into it.

--- a/module/interface_module_partitions/std.ccm
+++ b/module/interface_module_partitions/std.ccm
@@ -730,11 +730,11 @@ export
   } // namespace std
 
 
-// When compiling with Clang but using (as is the default) the GCC standard
-// library, we have to export a few symbols from the __gnu_cxx namespace as
-// well. Interestingly, we only need to do this when compiling with Clang but
-// not with GCC; the latter apparently finds these symbols itself and will
-// instead issue errors if we try to export them here.
+// When compiling with Clang but using libstdc++, export the libstdc++
+// iterator comparison operators from __gnu_cxx as well. With newer
+// libstdc++ versions these comparisons are available via operator<=>, but
+// Clang does not reliably synthesize the rewritten relational operators when
+// the iterators are used through an imported module interface.
 #if defined(__GLIBCXX__) && defined(__clang__)
 
   namespace __gnu_cxx
@@ -742,6 +742,51 @@ export
     using __gnu_cxx::operator-;
     using __gnu_cxx::operator==;
     using __gnu_cxx::operator<=>;
+
+    template <typename _IteratorL, typename _IteratorR, typename _Container>
+    [[nodiscard]] constexpr bool
+    operator<(const __normal_iterator<_IteratorL, _Container> &__lhs,
+              const __normal_iterator<_IteratorR, _Container>
+                &__rhs) noexcept(noexcept(__lhs.base() < __rhs.base()))
+    {
+      return __lhs.base() < __rhs.base();
+    }
+
+    template <typename _IteratorL, typename _IteratorR, typename _Container>
+    [[nodiscard]] constexpr bool
+    operator>(const __normal_iterator<_IteratorL, _Container> &__lhs,
+              const __normal_iterator<_IteratorR, _Container>
+                &__rhs) noexcept(noexcept(__rhs < __lhs))
+    {
+      return __rhs < __lhs;
+    }
+
+    template <typename _IteratorL, typename _IteratorR, typename _Container>
+    [[nodiscard]] constexpr bool
+    operator<=(const __normal_iterator<_IteratorL, _Container> &__lhs,
+               const __normal_iterator<_IteratorR, _Container>
+                 &__rhs) noexcept(noexcept(!(__rhs < __lhs)))
+    {
+      return !(__rhs < __lhs);
+    }
+
+    template <typename _IteratorL, typename _IteratorR, typename _Container>
+    [[nodiscard]] constexpr bool
+    operator>=(const __normal_iterator<_IteratorL, _Container> &__lhs,
+               const __normal_iterator<_IteratorR, _Container>
+                 &__rhs) noexcept(noexcept(!(__lhs < __rhs)))
+    {
+      return !(__lhs < __rhs);
+    }
+
+    template <typename _IteratorL, typename _IteratorR, typename _Container>
+    [[nodiscard]] constexpr bool
+    operator!=(const __normal_iterator<_IteratorL, _Container> &__lhs,
+               const __normal_iterator<_IteratorR, _Container>
+                 &__rhs) noexcept(noexcept(!(__lhs == __rhs)))
+    {
+      return !(__lhs == __rhs);
+    }
   } // namespace __gnu_cxx
 
 #elif defined(_LIBCPP_VERSION) // For CLang's support library

